src/pathfinder/yapf/yapf_costcache.hpp

```diff
@@ -123,27 +122,26 @@ struct CSegmentCostCacheT : public CSegmentCostCacheBase {
 	static const int C_HASH_BITS = 14;
 
 	typedef CHashTableT<Tsegment, C_HASH_BITS> HashTable;
-	typedef SmallArray<Tsegment> Heap;
 	typedef typename Tsegment::Key Key;    ///< key to hash table
 
 	HashTable    m_map;
-	Heap         m_heap;
+	std::vector<Tsegment> m_heap;
 
 	inline CSegmentCostCacheT() {}
 
 	/** flush (clear) the cache */
 	inline void Flush()
 	{
 		m_map.Clear();
-		m_heap.Clear();
+		m_heap.clear();
 	}
 
 	inline Tsegment& Get(Key &key, bool *found)
 	{
 		Tsegment *item = m_map.Find(key);
 		if (item == nullptr) {
 			*found = false;
-			item = new (m_heap.Append()) Tsegment(key);
+			item = &m_heap.emplace_back(key);
```

Comment:
It looks to me like SmallArray actually has stable iterators, so the pointer-to-member do not change when growing.
vector does not have that property.
I think we need std::deque for that.

Reply:
I haven't looked at the code, but just as a note for deque: `In addition, insertion and deletion at either end of a deque never invalidates pointers or references to the rest of the elements.`

Inserting/removing something in the middle will still invalidate pointers.